dot/state/epoch.go

```diff
@@ -365,3 +454,143 @@ func (s *EpochState) SkipVerify(header *types.Header) (bool, error) {
 
 	return false, nil
 }
+
+// StoreBABENextEpochData stores the types.NextEpochData under epoch and hash keys
+func (s *EpochState) StoreBABENextEpochData(epoch uint64, hash common.Hash, val types.NextEpochData) {
+	s.nextEpochLock.Lock()
+	defer s.nextEpochLock.Unlock()
+
+	_, has := s.nextEpochData[epoch]
+	if !has {
+		s.nextEpochData[epoch] = make(map[common.Hash]types.NextEpochData)
+	}
+	s.nextEpochData[epoch][hash] = val
+}
+
+// StoreBABENextConfigData stores the types.NextConfigData under epoch and hash keys
+func (s *EpochState) StoreBABENextConfigData(epoch uint64, hash common.Hash, val types.NextConfigData) {
+	s.nextEpochLock.Lock()
+	defer s.nextEpochLock.Unlock()
+
+	_, has := s.nextConfigData[epoch]
+	if !has {
+		s.nextConfigData[epoch] = make(map[common.Hash]types.NextConfigData)
+	}
+	s.nextConfigData[epoch][hash] = val
+}
+
+// FinalizeBABENextEpochData stores the right types.NextEpochData by
+// getting the set of hashes from the received epoch and for each hash
+// check if the header is in the database then it's been finalized and
+// thus we can also set the corresponding EpochData in the database
+func (s *EpochState) FinalizeBABENextEpochData(epoch uint64) error {
+	s.nextEpochLock.Lock()
+	defer s.nextEpochLock.Unlock()
+
+	finalizedNextEpochData, err := s.lookForPersistedHashForEpochData(epoch)
+	if err != nil {
+		return fmt.Errorf("cannot find next epoch data: %w", err)
+	}
+
+	ed, err := finalizedNextEpochData.ToEpochData()
+	if err != nil {
+		return fmt.Errorf("cannot transform epoch data: %w", err)
+	}
+
+	err = s.SetEpochData(epoch, ed)
+	if err != nil {
+		return fmt.Errorf("cannot set epoch data: %w", err)
+	}
+
+	// remove previous epochs from the memory
+	epochsToDelete := make([]uint64, 0)
```

Comment:
nit Let's assign it some approximate capacity

```suggestion
	epochsToDelete := make([]uint64, 0, len(s.nextEpochData)/2)
```